### DISCUSSION TITLE
MSC Read on slow data source

### DISCUSSION THREAD

**Discussion by Jubatian** [hathach/tinyusb#2034] [General]
Hi,

Would like to ask about the behaviour of **tud_msc_read10_cb()** in a scenario which seems to not work like I was expecting.

I have a relatively slow data source (an SPI Flash chip).

The existing architecture (with a different USB library I am migrating from), when a read request arrives from USB, calls a callback to notify the Data Source module that it needs to prepare "lba" sector.

The Data Source module later called from the main loop manages the read working through a state machine in a couple of main loop iterations until the data is ready. When so, it calls the USB library, which then begins sending the response.

Just to emphasize: There is no RTOS or interrupts in play, it is all sequential program flow.

Reading **tud_msc_read10_cb()**, it appeared like it should be easy, return 0 to it until the data is ready. However what I get is that seemingly it never returns! (The main loop freezes up for several seconds, with no return data observable by Wireshark).

Digging into TinyUSB's code I have a suspicion that it might not work the way I anticipated here (**tud_task()** doesn't return until **tud_msc_read10_cb()** is served a nonzero return, so there is no chance for other things called from the main loop to do their jobs, including preparing the data to be sent back).

Mainly would like to ask, whether I see this right? (To avoid spending more time on finding out why it doesn't work if this is it)

If it indeed works this way, any possible suggestions would be nice of course (maybe a way to tell data not yet ready to the host with an SCSI sense code? Not sure).

**Comment by hathach** [comment]
yeah, I think you are right about this. The way the msc driver is doing is queuing its self back using the same usbd event queue. Since tud_task() trying to process all the event before return. This cause an loop() without giving back the cpu. I think we could fix this by adding custom field for tud_task() to skip it until next time or so.

PS: I made an issue based on this discussion, since it is indeed an bug. Let move discussion there, thanks.
